packages/api-codec/src/StorageKey.ts

```diff
@@ -2,8 +2,67 @@
 // This software may be modified and distributed under the terms
 // of the ISC license. See the LICENSE file for details.
 
+import { AnyU8a } from './types';
+
+import isFunction from '@polkadot/util/is/function';
+
 import Bytes from './Bytes';
+import { StorageFunctionMetadata } from './Metadata';
+
+export interface StorageFunction extends StorageFunctionMetadata {
+  create (arg?: any): StorageKey;
+}
 
-// A representation of a storage key (typically hashed) in the system
+export type StorageFunctionWithArg = [StorageFunction, any];
+
+function isStorageFunction (value?: any): value is StorageFunction {
+  return value && isFunction(value.create);
+}
+
+// A representation of a storage key (typically hashed) in the system. It can be constructed
+// by passing in a raw key or a StorageFunction with (optional) arguments.
 export default class StorageKey extends Bytes {
+  private _outputType: string | null;
+
+  constructor (value: AnyU8a | StorageKey | StorageFunction | StorageFunctionWithArg = new Uint8Array(), ...args: Array<any>) {
```

Comment:
Not sure we should actually allow AnyU8a, it is useful, but if we funnel towards using StorageFunction as a constant, probably not the best idea to have too many options.